ledger/src/blocktree.rs

```diff
@@ -1611,6 +1621,37 @@ impl Blocktree {
         self.dead_slots_cf.put(slot, &true)
     }
 
+    pub fn store_duplicate_slot(&self, slot: Slot, shred1: Vec<u8>, shred2: Vec<u8>) -> Result<()> {
+        let duplicate_slot_proof = DuplicateSlotProof::new(shred1, shred2);
+        self.duplicate_slots_cf.put(slot, &duplicate_slot_proof)
+    }
+
+    pub fn get_duplicate_slot(&self, slot: u64) -> Option<DuplicateSlotProof> {
+        self.duplicate_slots_cf
+            .get(slot)
+            .expect("fetch from DuplicateSlots column family failed")
+    }
+
+    // `new_shred` is asssumed to have slot and index equal to the given slot and index.
+    // Returns true if `new_shred` is not equal to the existing shred at the given
+    // slot and index as this implies the leader generated two different shreds with
+    // the same slot and index
+    pub fn is_shred_duplicate(&self, slot: u64, index: u64, new_shred: &[u8]) -> bool {
+        let res = self
+            .get_data_shred(slot, index)
+            .expect("fetch from DuplicateSlots column family failed");
+
+        res.map(|existing_shred| existing_shred != new_shred)
+            .unwrap_or(false)
+    }
+
+    pub fn is_duplicate_slot(&self, slot: Slot) -> bool {
```

Comment:
This returns true if the node observed duplicate shreds for this slot, right? I think the naming could be better, maybe `has_duplicate_shreds_in_slot()`

Reply:
actually do we even need both this and `get_duplicate_slot`?

Reply:
hehe one is just a convenience method, you could also just do `get_duplicate_slot().is_some()`, you think thats sufficient?